pkg/server/config.go

```diff
@@ -335,23 +332,35 @@ func SetOpenFileLimitForOneStore() (uint64, error) {
 	return setOpenFileLimit(1)
 }
 
-// MakeTempStoreSpecFromStoreSpec creates a spec for a temporary store under
-// the given StoreSpec's path. If the given spec specifies an in-memory store,
-// the temporary store will be in-memory as well. The Attributes field of the
-// given spec is intentionally not propagated to the temporary store.
-//
-// TODO(arjun): Add a CLI flag to override this.
-func MakeTempStoreSpecFromStoreSpec(spec base.StoreSpec) base.StoreSpec {
-	if spec.InMemory {
+// MakeTempStoreSpecFromStoreSpec creates a spec for a temporary store. If the desiredDir
+// is not specified with the --temp-dir flag, then a subdirectory under the firstStore's path
+// will be used for the temporary store.
+// The Attributes field of the given spec is intentionally not propagated to the temporary store.
+func MakeTempStoreSpecFromStoreSpec(
+	desiredDir string, firstStore base.StoreSpec,
+) (base.StoreSpec, error) {
+	if firstStore.InMemory {
 		return base.StoreSpec{
 			// TODO(arjun): Set the size in a principled fashion from the main store
 			// after #16750 is addressed.
 			InMemory: true,
-		}
+		}, nil
+	}
+
+	// No temporary directory specified in CLI flag:
+	// default temporary subdirectory under firstStore's path.
+	if desiredDir == "" {
+		desiredDir = firstStore.Path
 	}
-	return base.StoreSpec{
-		Path: filepath.Join(spec.Path, defaultTempStoreRelativePath),
+
+	// Guaranteed to always generate a unique temporary directory with the
+	// prefix "cockroach-temp".
+	tempPath, err := ioutil.TempDir(desiredDir, "cockroach-temp")
```

Comment:
As discussed with @knz briefly, this creates a temporary directory right away (compared to previously where no directories are created and only the filepath is obtained).